### PR TITLE
Fix: db string conversion

### PR DESCRIPTION
**Pull request by david-allison-1**
loading from TEXT -> Long required strtoll
loading from TEXT -> Double required strtod

Both of these are copied implementations

Related: https://github.com/ankidroid/Anki-Android/issues/8877

### CODE REVIEW

File: rsdroid/src/main/java/net/ankiweb/rsdroid/utils/StringToLong.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+        if (start == pos) {
+            return 0;
+        }
+
+        s = s.substring(start, pos);
+
+        try {
+            return Long.parseLong(s, radix);
+        } catch (NumberFormatException e) {
+            return Long.MAX_VALUE;
```
Comment (david-allison-1) [inline rsdroid/src/main/java/net/ankiweb/rsdroid/utils/StringToLong.java:78]: check

File: rsdroid-instrumented/src/androidTest/java/net/ankiweb/rsdroid/DatabaseIntegrationTests.java [hunk 2]
```diff
@@ -186,6 +188,8 @@ public void testIntConversions() {
         testIntConversion("double", 1); // this is a .floor
         testIntConversion("string", 0);
         testIntConversion("null", 0);
+        testIntFromStringConversion("2", 2);
+        testIntFromStringConversion("2.52", 2);
```
Comment (mikehardy) [inline rsdroid-instrumented/src/androidTest/java/net/ankiweb/rsdroid/DatabaseIntegrationTests.java:192]: In all of these, what about the error cases? Testing conversion from "IAmNotAnInt" to int / float / double etc, or just empty space, or null ? :thinking:

Comment (david-allison-1) [reply rsdroid-instrumented/src/androidTest/java/net/ankiweb/rsdroid/DatabaseIntegrationTests.java:192]: This is testing the error cases: 

https://github.com/david-allison-1/Anki-Android-Backend/blob/626d65076534653e4f3334c98070c12b49c4705d/rsdroid-instrumented/src/androidTest/java/net/ankiweb/rsdroid/DatabaseIntegrationTests.java#L421-L430

File: rsdroid/src/main/java/net/ankiweb/rsdroid/database/StreamingProtobufSQLiteCursor.java [hunk 3]
```diff
@@ -269,5 +269,22 @@ public int getType(int columnIndex) {
...
+        } catch (NumberFormatException exception) {
+            return 0;
+        }
+    }
+
+    private double strtod(String stringValue) {
+        try{
+
+        return StringToDouble.strtod(stringValue);
+        } catch (NumberFormatException exception) {
```
Comment (mikehardy) [inline rsdroid/src/main/java/net/ankiweb/rsdroid/database/StreamingProtobufSQLiteCursor.java:285]: spacing is off here

**Review by mikehardy — APPROVED** [review]
2 tiny things but nothing really substantive --> :+1:

File: rsdroid-instrumented/src/androidTest/java/net/ankiweb/rsdroid/DatabaseIntegrationTests.java (hunk 2 above)
Comment (david-allison-1) [reply rsdroid-instrumented/src/androidTest/java/net/ankiweb/rsdroid/DatabaseIntegrationTests.java:192]: I'll get this rewritten when I have time tonight - it's minor on the TODO list, but let's fix it now it's been flagged

**Comment by dae** [comment]
One alternative that might be worth considering is casting the strftime output at the sql level. This would make the conversion more explicit, avoid the need for pasted-in code, and prevent accidental usage of a string when a number was expected.

```
sqlite> select cast("1.6" as int);
cast("1.6" as int) = 1
sqlite> select cast("1.6" as float);
cast("1.6" as float) = 1.6
```

**Comment by mikehardy** [comment]
sqlite should run that stuff a great deal faster as well - good thought

**Comment by david-allison-1** [comment]
Added a separate issue

Avoiding the need for this conversion via SQL is the ideal, but we should honor the contract specified in `CursorWindow` regardless: 

> If the field is of type Cursor#FIELD_TYPE_STRING, then the result is the value obtained by parsing the string value with strtoll.

https://developer.android.com/reference/android/database/CursorWindow#getLong(int,%20int)

**Review by mikehardy — APPROVED** [review]
Tests are a lot more readable now

**Pull request merged**